enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,460 @@
+---
+title: on-prem-service-load-balancers
+authors:
+  - "@russellb"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-05-14
+last-updated: 2020-05-14
+status: proposed
+---
+
+# Service Load Balancers for On Premise Infrastructure
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+We do not currently support full automation for [Services of
+type=LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/#loadbalancer)
+(Service Load Balancers, or SLBs) for OpenShift in bare metal environments.
+While bare metal clusters are of primary interest, we hope to find a solution
+that would apply to other on-premise environments that don’t have native load
+balancer capabilities available (a cluster on VMware, RHV, or OpenStack without
+Octavia, for example).
+
+## Motivation
+
+Service Load Balancers are a common way to expose applications on a cluster. We
+highly value clusters using on premise infrastructure, but do not support this
+feature in that context.  We aim to fill this gap with an optional range of
+capabilities.
+
+We do have a related feature, `AutoAssignCIDRs`, where OpenShift will
+automatically assign an ExternalIP for SLBs.  However, routing traffic to these
+IPs is still left up as an exercise to the administrator.  This enhancement
+offers an improvement where we can automate making these IP addresses
+reachable.  This method would remain an option for any clusters where the
+administrator would like to manage routing for external IPs in a completely
+custom manner.  For more information on the existing `IngressIPs` feature:
+
+* https://github.com/openshift/api/blob/master/config/v1/types_network.go#L99
+* https://github.com/openshift/openshift-docs/pull/21388
+
+### Goals
+
+A SLB solution must provide these high level features:
```

Comment:
Stupid question time!

This list of features does not mention ... load balancing? i.e. distributing incoming connections over a set of available endpoints?

I assume this means the actual load balancing is done by the service proxy provided by openshift-sdn or ovn-kubernetes. And the frontend IP for that load balancer is the clusterIP field. And MetaLB is all about allocating a "public" IP for a service and advertising that IP.

First point of confusion - if a connection for the MetalLB-allocated IP arrives on a node, how does it end up being accepted by a pod?

And then back to ... load-balancing - with BGP, is it just that multiple nodes advertise themselves as listening on the MetalLB IP, and the load-balancing across those multiple listeners depends on the router configuration. And with the L2 mode, only one of the listening nodes advertises itself?

I dunno, I think I'd appreciate some brief discussion of ... load-balancing? (kthxbi!)

Reply:
> Stupid question time!

Not stupid at all.  This is by far the #1 point of confusion I've seen in the MetalLB community so far.  The docs really need to be improved to help clarify what MetalLB does and does not do, which funny enough does not include load balancing.

> This list of features does not mention ... load balancing? i.e. distributing incoming connections over a set of available endpoints?
> 
> I assume this means the actual load balancing is done by the service proxy provided by openshift-sdn or ovn-kubernetes. And the frontend IP for that load balancer is the clusterIP field. And MetaLB is all about allocating a "public" IP for a service and advertising that IP.

You got it.

> 
> First point of confusion - if a connection for the MetalLB-allocated IP arrives on a node, how does it end up being accepted by a pod?

Apparently service proxy implementations must already be ready to accept traffic to a Node with a destination IP address set to the external IP address for a LoadBalancer Service.  All MetalLB has to do is implement some environment-appropriate method of getting the traffic to a Node.

> 
> And then back to ... load-balancing - with BGP, is it just that multiple nodes advertise themselves as listening on the MetalLB IP, and the load-balancing across those multiple listeners depends on the router configuration. 

That is correct!  This is where "ECMP" comes in and we get real load balancing across Nodes for a single LoadBalancer.  You are correct that this relies on configuration of your router(s).

There's another detail to this ... there is a field called `externalTrafficPolicy` that can be `cluster` or `local`.   https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip

* `local` -- In this mode, the pod expects to receive traffic with the original source IP address still intact.  To achieve this, the traffic must go directly to a Node where one of the Endpoints for that Service is running.  Only those Nodes advertise the Service IP in this case.

* `cluster` -- In this mode, traffic may arrive on any Node and will be redirected to another Node if necessary to reach a Service Endpoint.  The source IP address will be changed to the Node's IP to ensure traffic returns via the same path it arrived on.  The Service IP is advertised for *all* Nodes for these Services.

> And with the L2 mode, only one of the listening nodes advertises itself?

That is correct.  There's an algorithm used inside MetalLB that ensures that all Nodes agree which one it should be.
 
> I dunno, I think I'd appreciate some brief discussion of ... load-balancing? (kthxbi!)

ha, OK.  Yes, I'll take a fresh look at this doc and add some more information about it.  I've learned more since I first wrote it.

Reply:
I pushed another commit with some additional information on these topics.  I hope it helps.

Reply:
Beautiful, that's super helpful!